cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,454 @@
+It's not possible to reuse an identity-identity handshake across nodes without making bad-node attacks much easier.
+
+
+### Control Plane
+
+#### Structures
+
+Auth table - a BPF table needs to be added that records, at least:
+
+* Local Identity
+* Remote Identity
+* Remote Node IP
+* Authentication expiry time
+
+Further details of this table will be worked out as part of the implementation.
+
+#### Connections
+
+TODO: This section needs updating
+
+##### Between Cilium Operator instances and the SPIFFE server
+
+SPIFFE identity registration must occur out-of-band from when the actual SPIFFE identity usage is made via the workload API. In order to register the SPIFFE identities, the Cilium Operator must share access to an admin domain socket with the SPIRE Server, and use that domain socket for creation of new SPIFFE identities in the SPIRE server.
+
+Notably, this means that _one_ Pod out of the Operator deployment must be running on the _same node_ as the SPIRE Server's StatefulSet Pod. This will need Pod Affinity, and some way for the Operator to check if the domain socket is available. This process will also need to run independently of the usual Kubernetes Leadership, unfortunately.
+
+##### Between Cilium Agent instance peers
+Mutual authentication connections are established between the node IPs themselves. However, the peers that require authentication will not be these nodes.
+
+The source agent determines the destination agent using the destination endpoint IP. It then reaches out and connects to the listening port on the destination agent, and completes a TLS handshake.
+After the TLS handshake is complete, the session is torn down, as all the relevant information has been required. (Namely, did the authentication work, and when do the client and server expire?)
+For the destination agent, this process is reversed (an incoming connection from identity A to B _from_ node A will create an entry with A, B, and Node A on Node B, whereas on Node A, the entry will have A, B, and Node B.)
+
+This design has a few wins here:
+- The Authentication only requires that the TLS Handshake succeed, it does not require that we keep the connection open, so we will be making short lived connections for each authentication check, and this will most likely need to be to a new, dedicated port.
+- Because we will be controlling both ends of the TLS handshake, we can use any field we like in the certificates for authentication - SPIFFE certificates issued by SPIRE have the full SPIFFE ID URI as a SAN field, so we can use that for choosing both client and server certificates.
+However, because the TLS handshake can only use a hostname for SNI, we'll need to make a standard "hostname" that encodes the source and destination security identities, probably something like `<sourceidentity>.<destinationidentity>.<trustdomain>`.
```

Comment:
Only the destination ID would be needed as source identification will be done with the supplied peer certificate which is what is also done in direct app to app mTLS

Reply:
oh, of course, thanks. 🤦